### PR TITLE
feat(input,textarea,datepicker,chip-input): focus api

### PR DESCRIPTION
**Pull request by ben-girardet**
Add a `focus()` and `blur()` public API to

* ux-input
* ux-textarea
* ux-datepicker
* ux-chip-input

Compatible with the Aurelia built-in `focus` custom attribute.

### CODE REVIEW

File: app/src/focus.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+export interface StringTMap<T> { [key: string]: T; };
+export interface StringAnyMap extends StringTMap<any> {};
+export interface StringBooleanMap extends StringTMap<boolean> {};
+
+export class Focus {
```
Comment (bigopon) [inline app/src/focus.ts]: Can you help add some comment describing the purpose of this class?

File: design-decisions.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+* display the component in its *focused* state
+* if text can be entered, it must focus the input tag, so that soft keyboard would open in mobile devices
+
+Also, if the inner HTML component receives focus (for exemple by the user hitting <TAB> key), the component must become in focused state.
+
+Since Aurelia has a built-in `focus` custom attribute, we expect Aurelia UX components to work seemlessly with this attribute. This constraints prevents us to use a `@bindable` for the focus attribute because the way Aurelia handles custom attributes conflicts with the bindable. The chosen solution is to add the `focus` and `blur` methods to the component's HTML element (through the custom prototype).
+
+### CSS
+
+When a component is in `focused` state, it must receive the CSS class `<comp-name>--focused`.
```
Comment (bigopon) [inline design-decisions.md]: pinging @EisenbergEffect @serifine for this change. At the moment, we are using `:focus` selector

File: app/src/focus.ts (hunk 1 above)
Comment (ben-girardet) [reply app/src/focus.ts]: Hum... 

`StringAnyMap` and `StringBooleanMap` are only interfaces that I often use when I need typings in objects. They should not be exported.

File: design-decisions.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+* `ux-input` (done)
+* `ux-textarea` (done)
+* `ux-datepicker` (done)
+* `ux-chip-input` (done)
+
+### Components with input or action on click
+
+*I am not sure if these components need to fully implement the focus() and blur() api. They do need to respond to <tab> keys and show a focused state. It seems to me, however, that they don't need to implement the focus() and blur() API as it doesn't make sense that we programatically want to focus a button or a checkbox?*
+
+* `ux-select` (tab works correctly)
```
Comment (bigopon) [inline design-decisions.md]: `ux-select` is a beast, in term of focus implementation, I think we can delay it if not critical

File: packages/chip-input/src/ux-chip-input.ts [hunk 4]
```diff
@@ -181,4 +186,76 @@ export class UxChipInput implements UxComponent {
 
     this.styleEngine.applyTheme(newValue, this.element);
   }
+
+  public focusedChanged(focused: boolean) {
+    if (focused === true) {
```
Comment (bigopon) [inline packages/chip-input/src/ux-chip-input.ts]: this is a good example of `toggle` api:
```html
this.element.classList.toggle('ux-chip-input--focused', focused);
```

File: design-decisions.md (hunk 2 above)
Comment (ben-girardet) [reply design-decisions.md]: Would also like to have their input on this one. It seems that at the moment there is a mix of both `:focus` and `<comp-name>--focused`. I was thinking that we are trying to use more BEM conventions, hence this suggestion.

File: packages/chip-input/src/ux-chip-input.ts [hunk 5]
```diff
@@ -181,4 +186,76 @@ export class UxChipInput implements UxComponent {
...
+  }
+
+  public focus() {
+    this.textbox.focus();
+  }
+
+  public blur() {
+    if (document.activeElement === this.textbox) {
+      this.textbox.blur();
+      this.addChip();
```
Comment (bigopon) [inline packages/chip-input/src/ux-chip-input.ts]: adding chip on blur is quite a surprise, do we want this? at least it needs to be explicitly configured

File: packages/input/src/ux-input.ts [hunk 6]
```diff
@@ -202,23 +202,38 @@ export class UxInput implements UxComponent {
...
+  }
 }
 
 function stopEvent(e: Event) {
   e.stopPropagation();
 }
 
 const getVm = <T>(_: any) => _.au.controller.viewModel as T;
-const uxInputElementProto = Object.create(HTMLElement.prototype, {
+const uxInputElementProto = Object.assign(
```
Comment (bigopon) [inline packages/input/src/ux-input.ts]: no need for `Object.assign`

File: packages/textarea/src/ux-textarea.css [hunk 7]
```diff
@@ -34,7 +34,7 @@
 .ux-textarea__bottom-border {
   height: 2px;
   background-color: #4043ff;
-  background-color: var(--aurelia-ux--chip-input-border-bottom-focus, var(--aurelia-ux--design-primary, #4043ff));
+  background-color: var(--aurelia-ux--textarea-border-bottom-focus, var(--aurelia-ux--design-primary, #4043ff));
```
Comment (bigopon) [inline packages/textarea/src/ux-textarea.css]: 👍

File: packages/textarea/src/ux-textarea.ts [hunk 8]
```diff
@@ -125,6 +126,11 @@ export class UxTextArea implements UxComponent {
   }
 
   public rawValueChanged(rawValue: string) {
+    if (rawValue.length > 0) {
+      this.element.classList.add('ux-textarea--has-value');
```
Comment (bigopon) [inline packages/textarea/src/ux-textarea.ts]: class list toggle

File: packages/chip-input/src/ux-chip-input.ts (hunk 5 above)
Comment (ben-girardet) [reply packages/chip-input/src/ux-chip-input.ts]: took this behavior from the previous blur listener (see detached() method on previous commit). For having used this component already in several projects I must say that I find it very practical from an end-user point of view.

When you say that it must be explicitly configured, do you think about a bindable to configure this as a component instance scope or from the plugin configure api to set it for all ux-chip-input in a project ?

Comment (bigopon) [reply packages/chip-input/src/ux-chip-input.ts]: I see. yes it should be a bindable where we can toggle this behavior. Thanks for pointing out the existing behavior

Comment (ben-girardet) [reply packages/chip-input/src/ux-chip-input.ts]: Will do.

As I like the `addChipOnBlur` behavior I would go for a default to `true` and allow user to opt-out. However, from your original remark you might suggest to have it `false` as default and require user to explicitly opt-in for it. Your thoughts ?

File: design-decisions.md (hunk 3 above)
Comment (ben-girardet) [reply design-decisions.md]: This PR leaves the `ux-select` as is. I agree with you

File: packages/input/src/ux-input.ts (hunk 6 above)
Comment (ben-girardet) [reply packages/input/src/ux-input.ts]: Are you sure ? the focus() and blur() are added outside the Object.create() call. How would it look like without the `Object.assign` ?

File: packages/chip-input/src/ux-chip-input.ts (hunk 5 above)
Comment (bigopon) [reply packages/chip-input/src/ux-chip-input.ts]: Well, ux is used in prod, so we cannot flip it around. maybe default to `true` and have option to `false` it

File: packages/input/src/ux-input.ts (hunk 6 above)
Comment (bigopon) [reply packages/input/src/ux-input.ts]: Oh I see, is there any reason you don't put it inside the `Object.create` call?
```ts
Object.create(HTMLElement.prototype, {
  value: {
    get() {
      return getVm<UxInput>(this).getValue();
    },
    set(value: any) {
      getVm<UxInput>(this).setValue(value);
    }
  },
  focus: {
    value() {
      getVm<UxInput>(this).focused = true;
    }
  }
}),
```

Comment (ben-girardet) [reply packages/input/src/ux-input.ts]: No reason, in fact I copied the code I saw in the related issue. Thanks for your exemple, I now removed `Object.assign()` in input, textarea, datepicker and chip-input.

File: app/src/focus.ts (hunk 1 above)
Comment (ben-girardet) [reply app/src/focus.ts]: @bigopon maybe you were talking about the `Focus` class ? It's only the view-model for the `focus` custom element in the test app. I only wanted to have a dedicated code source to test the focus attribute and create a simple custom element for this.

**Comment by ben-girardet** [comment]
@bigopon what's the status on this PR from your point of view ?

**Comment by bigopon** [comment]
@ben-girardet as for the direction, I think it's good, as it enhances the custom element hosts to have similar API to native inputs. For the details of the focus behavior, I dont think we will have clear answer to all edge and corner cases, better let them with some consistency and then later draft out a better specs if we see any issues. For now, I think focusing on the underlying textbox is good enough. Let me know when you need a review for the code, as we don't have tests or specs, it will feel uneasy, but it should be ok if we continue the course of this until that point.

**Comment by ben-girardet** [comment]
@bigopon You can go ahead with a code review. As you mention, it probably doesn't cover all cases, but from my experience (already using this code) it already greatly improve the focus api.

**Pull request closed without merging**